File: activity/marketplace.mjs
import { createSign } from "node:crypto";

const REPOSITORY_PATTERN = /^[A-Za-z0-9][A-Za-z0-9-]*\/[A-Za-z0-9._-]+$/;
const COMMIT_PATTERN = /^[0-9a-f]{40}$/i;
const SAFE_PATH_PATTERN = /^(?!\/)(?!.*(?:^|\/)\.\.(?:\/|$))[A-Za-z0-9._/-]*$/;
const INTERNAL_PACKAGES = new Set(["activity", "dashboard"]);
const MAX_MANIFEST_BYTES = 256 * 1024;
const MAX_PACKAGES_PER_REGISTRY = 500;

function base64url(value) {
  return Buffer.from(value).toString("base64url");
}

function appJwt(appId, privateKey, now = Date.now()) {
  const issuedAt = Math.floor(now / 1000) - 60;
  const header = base64url(JSON.stringify({ alg: "RS256", typ: "JWT" }));
  const payload = base64url(JSON.stringify({ iat: issuedAt, exp: issuedAt + 9 * 60, iss: appId }));
  const signingInput = `${header}.${payload}`;
  const signer = createSign("RSA-SHA256");
  signer.update(signingInput);
  signer.end();
  return `${signingInput}.${signer.sign(privateKey, "base64url")}`;
}

function secret(environment, name, field) {
  const value = typeof name === "string" ? environment[name] : "";
  if (!value) throw new Error(`${field} secret is unavailable`);
  return value;
}

function apiBase(registry) {
  return String(registry["api-url"] ?? "https://api.github.com").replace(/\/+$/, "");
}

async function githubRequest(fetchImpl, url, token, init = {}) {
  const response = await fetchImpl(url, {
    ...init,
    headers: {
      Accept: "application/vnd.github+json",
      "X-GitHub-Api-Version": "2022-11-28",
      ...(token ? { Authorization: ["Bearer", token].join(" ") } : {}),
      ...init.headers,
    },
  });
  if (!response.ok) throw new Error(`GitHub API request failed with status ${response.status}`);
  return response;
}

async function registryToken(registry, { fetchImpl, environment }) {
  const auth = registry.auth ?? { type: "none" };
  if (auth.type === "none") return "";
  if (auth.type === "pat") return secret(environment, auth.secret, "registry PAT");
  if (auth.type !== "github-app") throw new Error("registry authentication type is unsupported");
  const appId = secret(environment, auth["app-id-secret"], "registry GitHub App id");
  const privateKey = secret(environment, auth["private-key-secret"], "registry GitHub App private key");
  const installationId = secret(environment, auth["installation-id-secret"], "registry GitHub App installation id");
  const jwt = appJwt(appId, privateKey);
  const response = await githubRequest(
    fetchImpl,
    `${apiBase(registry)}/app/installations/${encodeURIComponent(installationId)}/access_tokens`,
    jwt,
    { method: "POST" },
  );
  const payload = await response.json();
  if (typeof payload.token !== "string" || !payload.token) {
    throw new Error("registry GitHub App token response is invalid");
  }
  return payload.token;
}

function scalar(source, name) {
  const value = source.match(new RegExp(`^${name}:[ \\t]*(.+?)\\s*$`, "m"))?.[1]?.trim() ?? "";
  if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
    return value.slice(1, -1);
  }
  return value;
}

function includes(source) {
  const block = source.match(/^includes:\s*\n((?:^[ \t]+.*\n?)*)/m)?.[1] ?? "";
  return block.split(/\r?\n/)
    .map((line) => line.match(/^\s*-\s+(?:source:\s+)?([^#]+?)\s*$/)?.[1])
    .filter(Boolean)
    .map((value) => value.replace(/^['"]|['"]$/g, ""));
}

export function parsePackageManifest(source, coordinates) {
  if (Buffer.byteLength(source) > MAX_MANIFEST_BYTES) throw new Error("package manifest exceeds size limit");
  const name = scalar(source, "name");
  if (!name) throw new Error("package manifest name is required");
  const contents = includes(source);
  const version = scalar(source, "version") || coordinates.ref;
  const packagePath = coordinates.path.replace(/\/?aw\.yml$/, "");
  const sourceCoordinate = `${coordinates.repository}${packagePath ? `/${packagePath}` : ""}@${coordinates.resolvedCommit}`;
  return {
    id: `${coordinates.registryId}:${sourceCoordinate}`,
    "registry-id": coordinates.registryId,
    "registry-name": coordinates.registryName,
    "registry-precedence": coordinates.precedence,
    name,
    description: scalar(source, "description"),
    publisher: coordinates.repository.split("/")[0],
    repository: coordinates.repository,
    path: packagePath,
    ref: coordinates.ref,
    "resolved-commit": coordinates.resolvedCommit,
    version,
    icon: scalar(source, "icon") || "workflow",
    artwork: scalar(source, "artwork") || "",
    contents,
    source: sourceCoordinate,
    "add-command": `./cao.sh add ${sourceCoordinate}`,
  };
}

function validateRegistry(registry, index) {
  if (!registry || typeof registry !== "object" || Array.isArray(registry)) throw new Error(`registry ${index} is invalid`);
  if (!/^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(registry.id ?? "")) throw new Error(`registry ${index} id is invalid`);
  if (!REPOSITORY_PATTERN.test(registry.repository ?? "")) throw new Error(`registry ${registry.id} repository is invalid`);
  if (typeof registry.ref !== "string" || !registry.ref.trim()) throw new Error(`registry ${registry.id} ref is required`);
  if (!SAFE_PATH_PATTERN.test(registry.path ?? "")) throw new Error(`registry ${registry.id} path is invalid`);
  return {
    ...registry,
    path: String(registry.path ?? "").replace(/^\/|\/$/g, ""),
    name: String(registry.name ?? registry.id),
  };
}

function safeDiagnostic(error, registry, environment) {
  let message = String(error?.message ?? error);
  for (const reference of Object.values(registry?.auth ?? {}).filter((value) => typeof value === "string")) {
    const value = environment[reference];
    if (typeof value === "string" && value) message = message.replaceAll(value, "[redacted]");
  }
  return message.slice(0, 500);
}

async function resolveRegistry(registry, precedence, options) {
  const token = await registryToken(registry, options);
  const base = apiBase(registry);
  const repositoryPath = registry.repository.split("/").map(encodeURIComponent).join("/");
  const commitResponse = await githubRequest(
    options.fetchImpl,
    `${base}/repos/${repositoryPath}/commits/${encodeURIComponent(registry.ref)}`,
    token,
  );
  const commitPayload = await commitResponse.json();
  const commit = String(commitPayload.sha ?? "");
  if (!COMMIT_PATTERN.test(commit)) throw new Error("registry ref did not resolve to a commit");
  const treeResponse = await githubRequest(
    options.fetchImpl,
    `${base}/repos/${repositoryPath}/git/trees/${commit}?recursive=1`,
    token,
  );
  const treePayload = await treeResponse.json();
  if (!Array.isArray(treePayload.tree) || treePayload.truncated === true) {
    throw new Error("registry tree is unavailable or truncated");
  }
  const prefix = registry.path ? `${registry.path}/` : "";
  const manifests = treePayload.tree
    .filter((entry) => entry?.type === "blob"
      && typeof entry.path === "string"
      && entry.path.startsWith(prefix)
      && entry.path.endsWith("/aw.yml")
      && entry.path !== `${prefix}aw.yml`)
    .filter((entry) => !INTERNAL_PACKAGES.has(entry.path.slice(prefix.length).split("/", 1)[0]))
    .slice(0, MAX_PACKAGES_PER_REGISTRY);
  return Promise.all(manifests.map(async (entry) => {
    const blobResponse = await githubRequest(options.fetchImpl, `${base}/repos/${repositoryPath}/git/blobs/${entry.sha}`, token);
    const blob = await blobResponse.json();
    if (blob.encoding !== "base64" || typeof blob.content !== "string") throw new Error("package manifest blob is invalid");
    const manifest = Buffer.from(blob.content.replace(/\s/g, ""), "base64").toString("utf8");
    if (scalar(manifest, "private") === "true") return null;
    return parsePackageManifest(manifest, {
      registryId: registry.id,
      registryName: registry.name,
      precedence,
      repository: registry.repository,
      path: entry.path,
      ref: registry.ref,
      resolvedCommit: commit,
    });
  })).then((packages) => packages.filter(Boolean));
}

export async function resolveMarketplace(marketplace, options = {}) {
  const registries = marketplace?.registries ?? [];
  if (!Array.isArray(registries)) throw new Error("marketplace registries must be a sequence");
  const fetchImpl = options.fetchImpl ?? globalThis.fetch;
  const environment = options.environment ?? process.env;
  const packages = [];
  const diagnostics = [];
  for (let index = 0; index < registries.length; index += 1) {
    let registry;
    try {
      registry = validateRegistry(registries[index], index);
      const resolved = await resolveRegistry(registry, index, { fetchImpl, environment });
      packages.push(...resolved);
      diagnostics.push({ "registry-id": registry.id, status: "available", packages: resolved.length });
    } catch (error) {
      diagnostics.push({
        "registry-id": registry?.id ?? String(registries[index]?.id ?? `registry-${index}`),
        status: "unavailable",
        message: safeDiagnostic(error, registry ?? registries[index], environment),
      });
    }
  }
  const seen = new Set();
  const ordered = packages
    .sort((left, right) => left["registry-precedence"] - right["registry-precedence"]
      || left.repository.localeCompare(right.repository)
      || left.path.localeCompare(right.path))
    .filter((entry) => {
      const coordinate = `${entry.repository}/${entry.path}`.toLowerCase();
      if (seen.has(coordinate)) return false;
      seen.add(coordinate);
      return true;
    });
  return { packages: ordered, diagnostics };
}

File: server/internal/marketplace/marketplace.go
// Package marketplace resolves the read-only campaign package catalog
// described in specs/marketplace.md for the hosted Go dashboard backend.
//
// It mirrors activity/marketplace.mjs so both backends return the same safe
// package DTO from the same ordered registries: registries are read from the
// CAO policy (control-plane.marketplace), each is resolved independently so
// one registry's failure never affects another, and normalized results are
// safe to cache and to serve to dashboard clients because they never carry
// secret values or access tokens.
package marketplace

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/githubnext/gh-aw-cao/server/internal/model"
)

// SourceName is the canonical dashboard source name the query engine and
// generation loader use to request the marketplace catalog.
const SourceName = "marketplace-packages"

// DefaultCacheTTL applies when the policy omits or sets an invalid
// cache-ttl-seconds value.
const DefaultCacheTTL = 5 * time.Minute

const (
	maxPolicyBytes         = 4 << 20 // 4 MiB: cao.json is small; this bounds a misconfigured or hostile policy file.
	maxManifestBytes       = 256 * 1024
	maxPackagesPerRegistry = 500
)

// internalPackages are catalog-owned directories that are never themselves
// installable packages.
var internalPackages = map[string]bool{"activity": true, "dashboard": true}

// AuthType enumerates the supported registry authentication modes.
type AuthType string

const (
	AuthNone      AuthType = "none"
	AuthPAT       AuthType = "pat"
	AuthGitHubApp AuthType = "github-app"
)

// Auth describes how a registry authenticates. Every field below other than
// Type is an environment variable *name*, never a secret value: resolvers
// read the named variable from the trusted hosted backend's own environment.
type Auth struct {
	Type                 AuthType `json:"type"`
	Secret               string   `json:"secret,omitempty"`
	AppIDSecret          string   `json:"app-id-secret,omitempty"`
	PrivateKeySecret     string   `json:"private-key-secret,omitempty"`
	InstallationIDSecret string   `json:"installation-id-secret,omitempty"`
}

// Registry is one ordered marketplace source. Earlier registries have higher
// precedence when two registries expose the same package coordinate.
type Registry struct {
	ID         string `json:"id"`
	Name       string `json:"name,omitempty"`
	Repository string `json:"repository"`
	Path       string `json:"path,omitempty"`
	Ref        string `json:"ref"`
	APIURL     string `json:"api-url,omitempty"`
	Auth       Auth   `json:"auth,omitempty"`
}

// Config is the parsed control-plane.marketplace policy section.
type Config struct {
	CacheTTL   time.Duration
	Registries []Registry
}

// Package is the normalized, safe package DTO shared with activity/marketplace.mjs.
type Package struct {
	ID                 string   `json:"id"`
	RegistryID         string   `json:"registry-id"`
	RegistryName       string   `json:"registry-name"`
	RegistryPrecedence int      `json:"registry-precedence"`
	Name               string   `json:"name"`
	Description        string   `json:"description"`
	Publisher          string   `json:"publisher"`
	Repository         string   `json:"repository"`
	Path               string   `json:"path"`
	Ref                string   `json:"ref"`
	ResolvedCommit     string   `json:"resolved-commit"`
	Version            string   `json:"version"`
	Icon               string   `json:"icon"`
	Artwork            string   `json:"artwork"`
	Contents           []string `json:"contents"`
	Source             string   `json:"source"`
	AddCommand         string   `json:"add-command"`
}

// Row converts the package to a dashboard row. Field names are kebab-case to
// match the canonical data model used by every other source.
func (p Package) Row() model.Row {
	contents := p.Contents
	if contents == nil {
		contents = []string{}
	}
	return model.Row{
		"id":                  p.ID,
		"registry-id":         p.RegistryID,
		"registry-name":       p.RegistryName,
		"registry-precedence": p.RegistryPrecedence,
		"package-name":        p.Name,
		"package-description": p.Description,
		"publisher":           p.Publisher,
		"repository":          p.Repository,
		"path":                p.Path,
		"package-ref":         p.Ref,
		"resolved-commit":     p.ResolvedCommit,
		"package-version":     p.Version,
		"package-icon":        p.Icon,
		"package-artwork":     p.Artwork,
		"package-contents":    contents,
		"package-source":      p.Source,
		"add-command":         p.AddCommand,
	}
}

// RegistryDiagnostic reports one registry's resolution outcome without any
// credential-derived detail.
type RegistryDiagnostic struct {
	RegistryID string `json:"registry-id"`
	Status     string `json:"status"`
	Packages   int    `json:"packages,omitempty"`
	Message    string `json:"message,omitempty"`
}

// Row converts the diagnostic to a dashboard-safe row.
func (d RegistryDiagnostic) Row() model.Row {
	row := model.Row{"registry-id": d.RegistryID, "status": d.Status}
	if d.Status == "available" {
		row["packages"] = d.Packages
	}
	if d.Message != "" {
		row["message"] = d.Message
	}
	return row
}

// Result is the outcome of resolving every configured registry.
type Result struct {
	Packages    []Package
	Diagnostics []RegistryDiagnostic
}

type policyDocument struct {
	ControlPlane struct {
		Marketplace *marketplaceSection `json:"marketplace"`
	} `json:"control-plane"`
}

type marketplaceSection struct {
	CacheTTLSeconds int        `json:"cache-ttl-seconds"`
	Registries      []Registry `json:"registries"`
}

// ParseConfig reads the control-plane.marketplace section out of a full CAO
// policy document. A policy without a marketplace section resolves to zero
// registries rather than an error, matching the catalog's opt-in schema.
func ParseConfig(data []byte) (*Config, error) {
	if len(data) > maxPolicyBytes {
		return nil, fmt.Errorf("CAO policy exceeds %d bytes", maxPolicyBytes)
	}
	var document policyDocument
	if err := json.Unmarshal(data, &document); err != nil {
		return nil, fmt.Errorf("parse CAO policy: %w", err)
	}
	if document.ControlPlane.Marketplace == nil {
		return &Config{Registries: []Registry{}}, nil
	}
	section := document.ControlPlane.Marketplace
	ttl := time.Duration(section.CacheTTLSeconds) * time.Second
	registries := section.Registries
	if registries == nil {
		registries = []Registry{}
	}
	return &Config{CacheTTL: ttl, Registries: registries}, nil
}

// LoadConfigFile reads and parses the CAO policy at path.
func LoadConfigFile(path string) (*Config, error) {
	// #nosec G304 -- path is an operator-controlled policy location (a flag/env
	// default), not user input.
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("read CAO policy %q: %w", path, err)
	}
	return ParseConfig(data)
}

var (
	registryIDPattern = regexp.MustCompile(`^[a-z0-9]+(?:-[a-z0-9]+)*$`)
	repositoryPattern = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9-]*/[A-Za-z0-9._-]+$`)
	secretNamePattern = regexp.MustCompile(`^[A-Z][A-Z0-9_]{0,127}$`)
	safePathCharset   = regexp.MustCompile(`^[A-Za-z0-9._/-]*$`)
	commitPattern     = regexp.MustCompile(`(?i)^[0-9a-f]{40}$`)
)

// isSafePath rejects absolute paths, disallowed characters, and any ".."
// path segment. Go's RE2 engine has no lookahead, so this is checked in code
// instead of by a single regular expression as activity/marketplace.mjs does.
func isSafePath(value string) bool {
	if strings.HasPrefix(value, "/") {
		return false
	}
	if !safePathCharset.MatchString(value) {
		return false
	}
	for _, segment := range strings.Split(value, "/") {
		if segment == ".." {
			return false
		}
	}
	return true
}

// ValidateRegistry checks structural and authentication-reference safety and
// returns a normalized copy (trimmed path, defaulted name). It never reads
// secrets; it only checks that auth fields *look like* environment variable
// names.
func ValidateRegistry(registry Registry, index int) (Registry, error) {
	if !registryIDPattern.MatchString(registry.ID) {
		return Registry{}, fmt.Errorf("registry %d id is invalid", index)
	}
	if !repositoryPattern.MatchString(registry.Repository) {
		return Registry{}, fmt.Errorf("registry %s repository is invalid", registry.ID)
	}
	if strings.TrimSpace(registry.Ref) == "" {
		return Registry{}, fmt.Errorf("registry %s ref is required", registry.ID)
	}
	path := strings.Trim(registry.Path, "/")
	if !isSafePath(path) {
		return Registry{}, fmt.Errorf("registry %s path is invalid", registry.ID)
	}
	if registry.APIURL != "" && !strings.HasPrefix(registry.APIURL, "https://") {
		return Registry{}, fmt.Errorf("registry %s api-url must use https", registry.ID)
	}
	if err := validateAuth(registry.Auth); err != nil {
		return Registry{}, fmt.Errorf("registry %s auth: %w", registry.ID, err)
	}
	normalized := registry
	normalized.Path = path
	if normalized.Name == "" {
		normalized.Name = registry.ID
	}
	return normalized, nil
}

func validateAuth(auth Auth) error {
	switch auth.Type {
	case "", AuthNone:
		return nil
	case AuthPAT:
		if !secretNamePattern.MatchString(auth.Secret) {
			return fmt.Errorf("pat secret must reference an environment variable name")
		}
		return nil
	case AuthGitHubApp:
		for _, value := range []string{auth.AppIDSecret, auth.PrivateKeySecret, auth.InstallationIDSecret} {
			if !secretNamePattern.MatchString(value) {
				return fmt.Errorf("github-app secrets must reference environment variable names")
			}
		}
		return nil
	default:
		return fmt.Errorf("authentication type %q is unsupported", auth.Type)
	}
}

// apiBase returns the registry's GitHub REST API base URL with any trailing
// slash removed, defaulting to github.com.
func apiBase(registry Registry) string {
	base := registry.APIURL
	if base == "" {
		base = "https://api.github.com"
	}
	return strings.TrimRight(base, "/")
}

func fallbackRegistryID(registry Registry, index int) string {
	if registry.ID != "" {
		return registry.ID
	}
	return fmt.Sprintf("registry-%d", index)
}

// safeDiagnosticMessage renders err for a registry diagnostic with any secret
// *value* referenced by the registry's auth configuration redacted first,
// mirroring safeDiagnostic in activity/marketplace.mjs. This matters even
// though resolvers never intentionally return secrets: an underlying HTTP
// client or upstream API can otherwise echo a token/PAT/private-key back into
// an error message (e.g. via a URL or transport error), and diagnostics are
// exposed to dashboard clients.
func safeDiagnosticMessage(err error, registry Registry, opts Options) string {
	message := err.Error()
	env := opts.env()
	for _, name := range authSecretNames(registry.Auth) {
		if name == "" {
			continue
		}
		if value, ok := env(name); ok && value != "" {
			message = strings.ReplaceAll(message, value, "[redacted]")
		}
	}
	if len(message) > 500 {
		return message[:500]
	}
	return message
}

func authSecretNames(auth Auth) []string {
	return []string{auth.Secret, auth.AppIDSecret, auth.PrivateKeySecret, auth.InstallationIDSecret}
}

// Resolve resolves every configured registry in order, isolating failures
// into diagnostics, and returns the deduplicated, precedence-sorted package
// catalog. generation isolates the Cache by dashboard data revision so a
// cached result from one revision is never served under another. cache may be
// nil to disable caching.
func Resolve(ctx context.Context, config *Config, generation string, cache Cache, opts Options) *Result {
	if config == nil {
		return &Result{Packages: []Package{}, Diagnostics: []RegistryDiagnostic{}}
	}
	ttl := config.CacheTTL
	if ttl <= 0 {
		ttl = DefaultCacheTTL
	}
	var packages []Package
	diagnostics := make([]RegistryDiagnostic, 0, len(config.Registries))
	for index, raw := range config.Registries {
		registry, err := ValidateRegistry(raw, index)
		if err != nil {
			diagnostics = append(diagnostics, RegistryDiagnostic{
				RegistryID: fallbackRegistryID(raw, index), Status: "unavailable", Message: safeDiagnosticMessage(err, raw, opts),
			})
			continue
		}
		if cached, hit := loadCachedRegistry(ctx, cache, registry.ID, generation); hit {
			packages = append(packages, cached...)
			diagnostics = append(diagnostics, RegistryDiagnostic{RegistryID: registry.ID, Status: "available", Packages: len(cached)})
			continue
		}
		resolved, err := ResolveRegistry(ctx, registry, index, opts)
		if err != nil {
			diagnostics = append(diagnostics, RegistryDiagnostic{
				RegistryID: registry.ID, Status: "unavailable", Message: safeDiagnosticMessage(err, registry, opts),
			})
			continue
		}
		storeCachedRegistry(ctx, cache, registry.ID, generation, ttl, resolved)
		packages = append(packages, resolved...)
		diagnostics = append(diagnostics, RegistryDiagnostic{RegistryID: registry.ID, Status: "available", Packages: len(resolved)})
	}
	return &Result{Packages: sortAndDedupe(packages), Diagnostics: diagnostics}
}

func sortAndDedupe(packages []Package) []Package {
	sort.SliceStable(packages, func(i, j int) bool {
		left, right := packages[i], packages[j]
		if left.RegistryPrecedence != right.RegistryPrecedence {
			return left.RegistryPrecedence < right.RegistryPrecedence
		}
		if left.Repository != right.Repository {
			return left.Repository < right.Repository
		}
		return left.Path < right.Path
	})
	seen := map[string]bool{}
	ordered := make([]Package, 0, len(packages))
	for _, entry := range packages {
		coordinate := strings.ToLower(entry.Repository + "/" + entry.Path)
		if seen[coordinate] {
			continue
		}
		seen[coordinate] = true
		ordered = append(ordered, entry)
	}
	return ordered
}
